internal/venice-client-common/src/main/java/com/linkedin/venice/schema/SchemaUtils.java

```diff
@@ -100,4 +105,91 @@ public static GenericRecord createGenericRecord(Schema originalSchema) {
+   * @param schema the input partial update schema to be annotated.
+   * @return Annotated partial update schema.
+   */
+  public static Schema annotateStringMapInPartialUpdateSchema(Schema schema) {
+    // Create duplicate schema here in order not to create any side effect during annotation.
+    Schema replicatedSchema = AvroSchemaParseUtils.parseSchemaFromJSONStrictValidation(schema.toString());
+    if (replicatedSchema.getType().equals(RECORD)) {
+      for (Schema.Field field: replicatedSchema.getFields()) {
+        if (field.schema().isUnion()) {
+          for (Schema unionBranchSchema: field.schema().getTypes()) {
+            // Full update request for Map field.
+            if (unionBranchSchema.getType().equals(Schema.Type.MAP)) {
+              annotateMapSchema(unionBranchSchema);
+            } else if (unionBranchSchema.getType().equals(RECORD)) {
+              for (Schema.Field updateOpField: unionBranchSchema.getFields()) {
+                Schema.Type updateOpFieldType = updateOpField.schema().getType();
+                if (updateOpFieldType.equals(MAP)) {
+                  annotateMapSchema(updateOpField.schema());
+                } else if (updateOpFieldType.equals(ARRAY)
+                    && updateOpField.schema().getElementType().getType().equals(STRING)) {
+                  annotateStringArraySchema(updateOpField.schema());
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    return replicatedSchema;
+  }
+
+  /**
+   * Create a new {@link SchemaEntry} for value schema annotation.
+   * @param schemaEntry Input {@link SchemaEntry}, containing the value schema to be annotated.
+   * @return Annotated value schema in a newly created {@link SchemaEntry}
+   */
+  public static SchemaEntry getAnnotatedStringMapValueSchemaEntry(SchemaEntry schemaEntry) {
+    if (schemaEntry == null) {
+      return null;
+    }
+    Schema annotatedSchema = annotateStringMapInValueSchema(schemaEntry.getSchema());
+    return new SchemaEntry(schemaEntry.getId(), annotatedSchema);
+  }
+
+  /**
+   * Create a new {@link DerivedSchemaEntry} for partial update schema annotation.
+   * @param schemaEntry Input {@link DerivedSchemaEntry}, containing the partial update schema to be annotated.
+   * @return Annotated partial update schema in a newly created {@link DerivedSchemaEntry}
+   */
+  public static DerivedSchemaEntry getAnnotatedStringMapDerivedSchemaEntry(DerivedSchemaEntry schemaEntry) {
+    if (schemaEntry == null) {
+      return null;
+    }
+    Schema annotatedSchema = annotateStringMapInPartialUpdateSchema(schemaEntry.getSchema());
+    return new DerivedSchemaEntry(schemaEntry.getValueSchemaID(), schemaEntry.getId(), annotatedSchema);
+  }
+
+  private static void annotateMapSchema(Schema mapSchema) {
+    AvroCompatibilityHelper.setSchemaPropFromJsonString(mapSchema, "avro.java.string", "\"String\"", false);
```

Comment:
I am wondering what will happen if the map/array already has the string annotation.
Can you double check?

Reply:
I just did a simple check on unit test, looks like it will have no effect on map that has string annotation.

Reply:
Cool, that is good to know!